## Feature Engieering (Cont.)

### Import Libraries

In [7]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from pylab import rcParams
rcParams["figure.figsize"] = (10, 6)
sns.set()
%matplotlib inline

### Import temp dataset

In [2]:
df2_ball = pd.read_csv("../Datasets/temp_df.csv")

In [3]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  
0  Royal Challengers Bangalore  
1  Royal Challengers Bangalore  
2  Royal Challengers Bangalore  
3  Royal Challengers Bangalore  
4  Royal Challengers Bangalore

### Create final_score column

In [4]:
df2_ball["final_score"] = 0

final_score = df2_ball.groupby(['id','inning']).total_runs.sum().reset_index()

for i in range(len(final_score)):
    temp = final_score.loc[i]
    df2_ball.loc[(df2_ball.id == temp.id) & (df2_ball.inning == temp.inning), "final_score"] = temp.total_runs

In [5]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  
0  Royal Challengers Bangalore          222  
1  Royal Challengers Bangalore          222  
2  Royal Challengers Bangalore          222  
3  Royal Challengers Bangalore          222  
4  Royal Challengers Bangalore          222

### Total runs, wickets and balls per over

In [6]:
df2_ball.tail()

id  inning  over  ball  total_runs  is_wicket    batting_team  \
193272  1237181       2  17.6     6           1          0  Mumbai Indians   
193273  1237181       2  18.1     1           1          0  Mumbai Indians   
193274  1237181       2  18.2     2           1          0  Mumbai Indians   
193275  1237181       2  18.3     3           0          1  Mumbai Indians   
193276  1237181       2  18.4     4           1          0  Mumbai Indians   

          bowling_team  final_score  
193272  Delhi Capitals          157  
193273  Delhi Capitals          157  
193274  Delhi Capitals          157  
193275  Delhi Capitals          157  
193276  Delhi Capitals          157

In [7]:
# is_wicket indicates wicket in particular over, we will create sum of total_wickets till particular over
# total_runs we will create sum of runs scored in particular over.
# ball we will create sum of balls delivered in particular over.

# Create new column for wickets and runs with initial value as 0
df2_ball['wickets'] = 0
df2_ball['runs'] = 0

# for ball we will start initial value with first ball delivered i.e 1
df2_ball['ball'] = 1

id = df2_ball.id.unique()

for i in id:
    for inning in range(1,3):
        index = df2_ball.loc[(df2_ball.id == i) & (df2_ball.inning == inning)].index
        for j in range(len(index)):
            current_index_df = df2_ball.loc[(df2_ball.index == index[j])]
            previous_index_df = df2_ball.loc[(df2_ball.index == index[j-1])]
            if j == 0:
                df2_ball.loc[df2_ball.index == index[j], "wickets"] = current_index_df.is_wicket.values[0]
                df2_ball.loc[df2_ball.index == index[j], "runs"] = current_index_df.total_runs.values[0]
                df2_ball.loc[df2_ball.index == index[j], "ball"] = current_index_df.ball.values[0]
            else:
                df2_ball.loc[df2_ball.index == index[j], "wickets"] = current_index_df.is_wicket.values[0] + previous_index_df.wickets.values[0]
                df2_ball.loc[df2_ball.index == index[j], "runs"] = current_index_df.total_runs.values[0] + previous_index_df.runs.values[0]
                df2_ball.loc[df2_ball.index == index[j], "ball"] = current_index_df.ball.values[0] + previous_index_df.ball.values[0]

In [8]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  wickets  runs  
0  Royal Challengers Bangalore          222        0     1  
1  Royal Challengers Bangalore          222        0     1  
2  Royal Challengers Bangalore          222        0     2  
3  Royal Challengers Bangalore          222        0     2  
4  Royal Challengers Bangalore          222        0     2

### Last 5 over's (runs, wickets, balls)

In [9]:
df2_ball.loc[df2_ball.over < int(str(current_index_df.over.values[0]).split(".")[0]) - 5]

id  inning  over  ball  total_runs  is_wicket  \
0        335982       1   0.1     1           1          0   
1        335982       1   0.2     2           0          0   
2        335982       1   0.3     3           1          0   
3        335982       1   0.4     4           0          0   
4        335982       1   0.5     5           0          0   
...         ...     ...   ...   ...         ...        ...   
193238  1237181       2  12.2    74           1          0   
193239  1237181       2  12.3    75           1          0   
193240  1237181       2  12.4    76           4          0   
193241  1237181       2  12.5    77           1          0   
193242  1237181       2  12.6    78           2          0   

                 batting_team                 bowling_team  final_score  \
0       Kolkata Knight Riders  Royal Challengers Bangalore          222   
1       Kolkata Knight Riders  Royal Challengers Bangalore          222   
2       Kolkata Knight Riders  Royal Challengers Bangalore          222   
3       Kolkata Knight Riders  Royal Challengers Bangalore          222   
4       Kolkata Knight Riders  Royal Challengers Bangalore          222   
...                       ...                          ...          ...   
193238         Mumbai Indians               Delhi Capitals          157   
193239         Mumbai Indians               Delhi Capitals          157   
193240         Mumbai Indians               Delhi Capitals          157   
193241         Mumbai Indians               Delhi Capitals          157   
193242         Mumbai Indians               Delhi Capitals          157   

        wickets  runs  
0             0     1  
1             0     1  
2             0     2  
3             0     2  
4             0     2  
...         ...   ...  
193238        2   104  
193239        2   105  
193240        2   109  
193241        2   110  
193242        2   112  

[130194 rows x 11 columns]

In [10]:
# let's create new column for last_5_over_wickets
df2_ball["last_5_over_wickets"] = 0
# let's create new column for last_5_over_runs
df2_ball["last_5_over_runs"] = 0
# let's create new column for last_5_over_balls
df2_ball["last_5_over_balls"] = 0
 # calculating last run, wickets and balls of last 5 over
for i in id:
    for inning in range(1, 3):
        index = df2_ball.loc[(df2_ball.id == i) & (df2_ball.inning == inning)].index
        for j in range(len(index)):
            current_index_df = df2_ball.loc[(df2_ball.index == index[j])]
            previous_index_df = df2_ball.loc[(df2_ball.index == index[j-1])]            
            if current_index_df.over.values[0] <= 5:
                df2_ball.loc[df2_ball.index == index[j], "last_5_over_wickets"] = current_index_df.wickets.values[0]
                df2_ball.loc[df2_ball.index == index[j], "last_5_over_runs"] = current_index_df.runs.values[0]
                df2_ball.loc[df2_ball.index == index[j], "last_5_over_balls"] = current_index_df.ball.values[0]
            else:
                df2_ball.loc[df2_ball.index == index[j], "last_5_over_wickets"] = current_index_df.wickets.values[0] - df2_ball.loc[(df2_ball.id == i) & (df2_ball.inning == inning) & (df2_ball.over < int(str(current_index_df.over.values[0]).split(".")[0]) - 5+1), "wickets"].values[0]
                df2_ball.loc[df2_ball.index == index[j], "last_5_over_runs"] = current_index_df.runs.values[0] - df2_ball.loc[(df2_ball.id == i) & (df2_ball.inning == inning) & (df2_ball.over < int(str(current_index_df.over.values[0]).split(".")[0]) - 5+1), "runs"].values[0]
                df2_ball.loc[df2_ball.index == index[j], "last_5_over_balls"] = current_index_df.ball.values[0] - df2_ball.loc[(df2_ball.id == i) & (df2_ball.inning == inning) & (df2_ball.over < int(str(current_index_df.over.values[0]).split(".")[0]) - 5+1), "ball"].values[0]

In [11]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  wickets  runs  \
0  Royal Challengers Bangalore          222        0     1   
1  Royal Challengers Bangalore          222        0     1   
2  Royal Challengers Bangalore          222        0     2   
3  Royal Challengers Bangalore          222        0     2   
4  Royal Challengers Bangalore          222        0     2   

   last_5_over_wickets  last_5_over_runs  last_5_over_balls  
0                    0                 1                  1  
1                    0                 1                  2  
2                    0                 2                  3  
3                    0                 2                  4  
4                    0                 2                  5

In [12]:
# Export the data frame
df2_ball.to_csv('../Datasets/df2_ball_featured_engineer_last5')

### Import match and df2_ball datasets

In [8]:
df2_ball = pd.read_csv("../Datasets/df2_ball_featured_engineer_last5.csv")
ipl_match = pd.read_csv("../Datasets/IPL Matches 2008-2020.csv")

In [9]:
ipl_match.head()

id        city        date player_of_match  \
0  335982   Bangalore  2008-04-18     BB McCullum   
1  335983  Chandigarh  2008-04-19      MEK Hussey   
2  335984       Delhi  2008-04-19     MF Maharoof   
3  335985      Mumbai  2008-04-20      MV Boucher   
4  335986     Kolkata  2008-04-20       DJ Hussey   

                                        venue  neutral_venue  \
0                       M Chinnaswamy Stadium              0   
1  Punjab Cricket Association Stadium, Mohali              0   
2                            Feroz Shah Kotla              0   
3                            Wankhede Stadium              0   
4                                Eden Gardens              0   

                         team1                        team2  \
0  Royal Challengers Bangalore        Kolkata Knight Riders   
1              Kings XI Punjab          Chennai Super Kings   
2             Delhi Daredevils             Rajasthan Royals   
3               Mumbai Indians  Royal Challengers Bangalore   
4        Kolkata Knight Riders              Deccan Chargers   

                   toss_winner toss_decision                       winner  \
0  Royal Challengers Bangalore         field        Kolkata Knight Riders   
1          Chennai Super Kings           bat          Chennai Super Kings   
2             Rajasthan Royals           bat             Delhi Daredevils   
3               Mumbai Indians           bat  Royal Challengers Bangalore   
4              Deccan Chargers           bat        Kolkata Knight Riders   

    result  result_margin eliminator method    umpire1         umpire2  
0     runs          140.0          N    NaN  Asad Rauf     RE Koertzen  
1     runs           33.0          N    NaN  MR Benson      SL Shastri  
2  wickets            9.0          N    NaN  Aleem Dar  GA Pratapkumar  
3  wickets            5.0          N    NaN   SJ Davis       DJ Harper  
4  wickets            5.0          N    NaN  BF Bowden     K Hariharan

In [10]:
df2_ball.head()

Unnamed: 0      id  inning  over  ball  total_runs  is_wicket  \
0           0  335982       1   0.1     1           1          0   
1           1  335982       1   0.2     2           0          0   
2           2  335982       1   0.3     3           1          0   
3           3  335982       1   0.4     4           0          0   
4           4  335982       1   0.5     5           0          0   

            batting_team                 bowling_team  final_score  wickets  \
0  Kolkata Knight Riders  Royal Challengers Bangalore          222        0   
1  Kolkata Knight Riders  Royal Challengers Bangalore          222        0   
2  Kolkata Knight Riders  Royal Challengers Bangalore          222        0   
3  Kolkata Knight Riders  Royal Challengers Bangalore          222        0   
4  Kolkata Knight Riders  Royal Challengers Bangalore          222        0   

   runs  last_5_over_wickets  last_5_over_runs  last_5_over_balls  
0     1                    0                 1                  1  
1     1                    0                 1                  2  
2     2                    0                 2                  3  
3     2                    0                 2                  4  
4     2                    0                 2                  5

In [11]:
df2_ball.drop(columns = 'Unnamed: 0',inplace=True)

In [12]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  wickets  runs  \
0  Royal Challengers Bangalore          222        0     1   
1  Royal Challengers Bangalore          222        0     1   
2  Royal Challengers Bangalore          222        0     2   
3  Royal Challengers Bangalore          222        0     2   
4  Royal Challengers Bangalore          222        0     2   

   last_5_over_wickets  last_5_over_runs  last_5_over_balls  
0                    0                 1                  1  
1                    0                 1                  2  
2                    0                 2                  3  
3                    0                 2                  4  
4                    0                 2                  5

In [13]:
ipl_match.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               816 non-null    int64  
 1   city             803 non-null    object 
 2   date             816 non-null    object 
 3   player_of_match  812 non-null    object 
 4   venue            816 non-null    object 
 5   neutral_venue    816 non-null    int64  
 6   team1            816 non-null    object 
 7   team2            816 non-null    object 
 8   toss_winner      816 non-null    object 
 9   toss_decision    816 non-null    object 
 10  winner           812 non-null    object 
 11  result           812 non-null    object 
 12  result_margin    799 non-null    float64
 13  eliminator       812 non-null    object 
 14  method           19 non-null     object 
 15  umpire1          816 non-null    object 
 16  umpire2          816 non-null    object 
dtypes: float64(1), i

In [16]:
df2_ball.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193277 entries, 0 to 193276
Data columns (total 16 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   id                   193277 non-null  int64  
 1   inning               193277 non-null  int64  
 2   over                 193277 non-null  float64
 3   ball                 193277 non-null  int64  
 4   total_runs           193277 non-null  int64  
 5   is_wicket            193277 non-null  int64  
 6   batting_team         193277 non-null  object 
 7   bowling_team         193277 non-null  object 
 8   final_score          193277 non-null  int64  
 9   wickets              193277 non-null  int64  
 10  runs                 193277 non-null  int64  
 11  last_5_over_wickets  193277 non-null  int64  
 12  last_5_over_runs     193277 non-null  int64  
 13  last_5_over_balls    193277 non-null  int64  
 14  venue                193277 non-null  object 
 15  winner           

In [17]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  wickets  runs  \
0  Royal Challengers Bangalore          222        0     1   
1  Royal Challengers Bangalore          222        0     1   
2  Royal Challengers Bangalore          222        0     2   
3  Royal Challengers Bangalore          222        0     2   
4  Royal Challengers Bangalore          222        0     2   

   last_5_over_wickets  last_5_over_runs  last_5_over_balls venue winner  
0                    0                 1                  1               
1                    0                 1                  2               
2                    0                 2                  3               
3                    0                 2                  4               
4                    0                 2                  5

### Add Venue and Winner Column

In [22]:
# Add winner and venue column from ipl_match
# Create column for venue in df2_ball
df2_ball["venue"] = ""
# let's create column for winner
df2_ball["winner"] = ""
for i in ipl_match.id:
    temp = ipl_match.loc[ipl_match.id == i]
    df2_ball.loc[df2_ball.id == i, "venue"] = temp.venue.values[0]
    df2_ball.loc[df2_ball.id == i, "winner"] = temp.winner.values[0]

In [23]:
df2_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  wickets  runs  \
0  Royal Challengers Bangalore          222        0     1   
1  Royal Challengers Bangalore          222        0     1   
2  Royal Challengers Bangalore          222        0     2   
3  Royal Challengers Bangalore          222        0     2   
4  Royal Challengers Bangalore          222        0     2   

   last_5_over_wickets  last_5_over_runs  last_5_over_balls  \
0                    0                 1                  1   
1                    0                 1                  2   
2                    0                 2                  3   
3                    0                 2                  4   
4                    0                 2                  5   

                   venue                 winner  
0  M Chinnaswamy Stadium  Kolkata Knight Riders  
1  M Chinnaswamy Stadium  Kolkata Knight Riders  
2  M Chinnaswamy Stadium  Kolkata Knight Riders  
3  M Chinnaswamy Stadium  Kolkata Knight Riders  
4  M Chinnaswamy Stadium  Kolkata Knight Riders

In [24]:
# Look for any null values in df2_ball
df2_ball.isnull().sum()

id                       0
inning                   0
over                     0
ball                     0
total_runs               0
is_wicket                0
batting_team             0
bowling_team             0
final_score              0
wickets                  0
runs                     0
last_5_over_wickets      0
last_5_over_runs         0
last_5_over_balls        0
venue                    0
winner                 181
dtype: int64

In [25]:
df2_ball[df2_ball.winner.isna()].id.unique()

array([ 829813, 1178424], dtype=int64)

In [26]:
df2_ball[df2_ball.winner.isna()]

id  inning  over  ball  total_runs  is_wicket      batting_team  \
120918   829813       1   0.1     1           1          0  Delhi Daredevils   
120919   829813       1   0.2     2           0          0  Delhi Daredevils   
120920   829813       1   0.3     3           1          0  Delhi Daredevils   
120921   829813       1   0.4     4           0          0  Delhi Daredevils   
120922   829813       1   0.5     5           0          0  Delhi Daredevils   
...         ...     ...   ...   ...         ...        ...               ...   
176154  1178424       2   2.4    16           0          0  Rajasthan Royals   
176155  1178424       2   2.5    17           6          0  Rajasthan Royals   
176156  1178424       2   2.6    18           4          0  Rajasthan Royals   
176157  1178424       2   3.1    19           1          0  Rajasthan Royals   
176158  1178424       2   3.2    20           0          1  Rajasthan Royals   

                       bowling_team  final_score  wickets  runs  \
120918  Royal Challengers Bangalore          187        0     1   
120919  Royal Challengers Bangalore          187        0     1   
120920  Royal Challengers Bangalore          187        0     2   
120921  Royal Challengers Bangalore          187        0     2   
120922  Royal Challengers Bangalore          187        0     2   
...                             ...          ...      ...   ...   
176154  Royal Challengers Bangalore           41        0    30   
176155  Royal Challengers Bangalore           41        0    36   
176156  Royal Challengers Bangalore           41        0    40   
176157  Royal Challengers Bangalore           41        0    41   
176158  Royal Challengers Bangalore           41        1    41   

        last_5_over_wickets  last_5_over_runs  last_5_over_balls  \
120918                    0                 1                  1   
120919                    0                 1                  2   
120920                    0                 2                  3   
120921                    0                 2                  4   
120922                    0                 2                  5   
...                     ...               ...                ...   
176154                    0                30                 16   
176155                    0                36                 17   
176156                    0                40                 18   
176157                    0                41                 19   
176158                    1                41                 20   

                        venue winner  
120918  M Chinnaswamy Stadium    NaN  
120919  M Chinnaswamy Stadium    NaN  
120920  M Chinnaswamy Stadium    NaN  
120921  M Chinnaswamy Stadium    NaN  
120922  M Chinnaswamy Stadium    NaN  
...                       ...    ...  
176154  M.Chinnaswamy Stadium    NaN  
176155  M.Chinnaswamy Stadium    NaN  
176156  M.Chinnaswamy Stadium    NaN  
176157  M.Chinnaswamy Stadium    NaN  
176158  M.Chinnaswamy Stadium    NaN  

[181 rows x 16 columns]

- [match id 829813](https://www.espncricinfo.com/series/pepsi-indian-premier-league-2015-791129/royal-challengers-bangalore-vs-delhi-daredevils-55th-match-829813/match-report)
- [match id 1178424](https://www.hindustantimes.com/cricket/rcb-vs-rr-live-score-ipl-2019-royal-challengers-bangalore-vs-rajasthan-royals-playing-11-match-49-at-chinnaswamy-stadium/story-fxdpeEBdw78wAFZYGomW4L.html)
- Both the matches were abandoned due to rain.
- Will drop the rows for this two match id's.

In [27]:
df2_ball.shape

(193277, 16)

In [28]:
# let's drop that NaN recorded matches and make a new data frame named df3_ball

df3_ball = df2_ball.dropna().reset_index(drop=True).copy()

In [30]:
df3_ball.head()

id  inning  over  ball  total_runs  is_wicket           batting_team  \
0  335982       1   0.1     1           1          0  Kolkata Knight Riders   
1  335982       1   0.2     2           0          0  Kolkata Knight Riders   
2  335982       1   0.3     3           1          0  Kolkata Knight Riders   
3  335982       1   0.4     4           0          0  Kolkata Knight Riders   
4  335982       1   0.5     5           0          0  Kolkata Knight Riders   

                  bowling_team  final_score  wickets  runs  \
0  Royal Challengers Bangalore          222        0     1   
1  Royal Challengers Bangalore          222        0     1   
2  Royal Challengers Bangalore          222        0     2   
3  Royal Challengers Bangalore          222        0     2   
4  Royal Challengers Bangalore          222        0     2   

   last_5_over_wickets  last_5_over_runs  last_5_over_balls  \
0                    0                 1                  1   
1                    0                 1                  2   
2                    0                 2                  3   
3                    0                 2                  4   
4                    0                 2                  5   

                   venue                 winner  
0  M Chinnaswamy Stadium  Kolkata Knight Riders  
1  M Chinnaswamy Stadium  Kolkata Knight Riders  
2  M Chinnaswamy Stadium  Kolkata Knight Riders  
3  M Chinnaswamy Stadium  Kolkata Knight Riders  
4  M Chinnaswamy Stadium  Kolkata Knight Riders

In [31]:
df3_ball.shape

(193096, 16)

### Venue feature 

In [32]:
# Venue feature

np.sort(df3_ball.venue.unique())

array(['Barabati Stadium', 'Brabourne Stadium', 'Buffalo Park',
       'De Beers Diamond Oval', 'Dr DY Patil Sports Academy',
       'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium',
       'Dubai International Cricket Stadium', 'Eden Gardens',
       'Feroz Shah Kotla', 'Green Park',
       'Himachal Pradesh Cricket Association Stadium',
       'Holkar Cricket Stadium', 'JSCA International Stadium Complex',
       'Kingsmead', 'M Chinnaswamy Stadium', 'M.Chinnaswamy Stadium',
       'MA Chidambaram Stadium, Chepauk',
       'Maharashtra Cricket Association Stadium', 'Nehru Stadium',
       'New Wanderers Stadium', 'Newlands', 'OUTsurance Oval',
       'Punjab Cricket Association IS Bindra Stadium, Mohali',
       'Punjab Cricket Association Stadium, Mohali',
       'Rajiv Gandhi International Stadium, Uppal',
       'Sardar Patel Stadium, Motera',
       'Saurashtra Cricket Association Stadium', 'Sawai Mansingh Stadium',
       'Shaheed Veer Narayan Singh International Stadium',


In [33]:
# Observed that M Chinnaswamy Stadium and M.Chinnaswamy Stadium both are same
# Punjab Cricket Association IS Bindra Stadium, Mohali and Punjab Cricket Association Stadium, Mohali both are same
# so let's make it one

df3_ball.venue.replace({"M.Chinnaswamy Stadium": "M Chinnaswamy Stadium", "Punjab Cricket Association IS Bindra Stadium, Mohali": "Punjab Cricket Association Stadium, Mohali"}, inplace=True)


In [34]:
# Export final featured engineering dataframe 

df3_ball.to_csv("eda_feature_engineering.csv", index=False)